src/plugins/intel_npu/src/plugin/npuw/weights_bank.hpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <mutex>
+#include <unordered_map>
+
+#include "openvino/openvino.hpp"
+#include "openvino/runtime/intel_npu/level_zero/level_zero.hpp"
+#include "openvino/runtime/iremote_context.hpp"
+#include "openvino/runtime/itensor.hpp"
+#include "openvino/runtime/make_tensor.hpp"
+#include "openvino/runtime/tensor.hpp"
+#include "plugin.hpp"
+
+namespace ov {
+namespace npuw {
+namespace weights_bank {
+
+class WeightBank {
+public:
+    WeightBank(const std::string& device_name, const std::shared_ptr<const ov::IPlugin>& plugin)
+        : m_device_name(device_name),
+          m_plugin(plugin) {
+        if (m_device_name != "CPU" && m_device_name != "NPU") {
+            OPENVINO_THROW("NPUW doesn't support ", m_device_name, " device for weights sharing!");
+        }
+        if (m_device_name == "NPU" && !m_remote_ctx) {
+            m_remote_ctx = m_plugin->get_core()->get_default_context("NPU")._ptr;
+        }
+    }
+
+    ov::SoPtr<ov::Tensor> getSharedTensor(const ov::element::Type& type, const ov::Shape& shape, void* host_data_ptr) {
+        if (!host_data_ptr) {
+            OPENVINO_THROW("Fatal: nullptr in weights bank allocation!");
+        }
+
+        std::lock_guard<std::mutex> guard(m_mutex);
+
+        // no special allocation needed
+        if (m_weights_bank.count(host_data_ptr) > 0) {
+            return m_weights_bank.at(host_data_ptr);
+        }
+
+        // need to allocate first
+        if (m_device_name == "CPU") {
+            m_weights_bank[host_data_ptr] = std::make_shared<ov::Tensor>(type, shape, host_data_ptr);
```

Comment:
if your API accepted tensors, you could just store one here.

Reply:
Done